### PR TITLE
Optimize performance of transaction-wise index query cache

### LINKED ISSUE #2194: Index query cache should support subset query
Within a transaction, JanusGraph stores index query results in a cache, where the key is JointIndexQuery.Subquery. This is not ideal for queries containing 'limit' step. See the following example:

```
gremlin> graph = JanusGraphFactory.open("conf/janusgraph-cql-es.properties")
14:55:32 WARN  com.datastax.driver.core.NettyUtil  - Found Netty's native epoll transport, but not running on linux-based operating system. Using NIO instead.
==>standardjanusgraph[cql:[127.0.0.1]]
gremlin>
gremlin> mgmt = graph.openManagement()
==>org.janusgraph.graphdb.database.management.ManagementSystem@3a209918
gremlin> nameKey = mgmt.makePropertyKey('name').dataType(String.class).make()
==>name
gremlin> mgmt.buildIndex("name", Vertex.class).addKey(nameKey, Mapping.STRING.asParameter()).buildMixedIndex("search")
==>name
gremlin> mgmt.commit()
==>null
gremlin>
gremlin> g = graph.traversal()
==>graphtraversalsource[standardjanusgraph[cql:[127.0.0.1]], standard]
gremlin> for (i = 0; i < 10000; i++) {
......1> g.addV().property("name", "value" + i % 10).iterate();
......2> }
==>null
gremlin> graph.tx().commit()
==>null
gremlin>
gremlin> g.V().has("name", "value0").limit(1000).profile()
==>Traversal Metrics
Step                                                               Count  Traversers       Time (ms)    % Dur
=============================================================================================================
JanusGraphStep([],[name.eq(value0)])                                                          18.195   100.00
    \_condition=(name = value0)
    \_orders=[]
    \_limit=1000
    \_isFitted=true
    \_isOrdered=true
    \_query=[(name = value0)](2000):name
    \_index=name
    \_index_impl=search
  optimization                                                                                 0.041
  optimization                                                                                 1.919
  backend-query                                                        0                      11.869
    \_query=name:[(name = value0)](2000):name
    \_limit=2000
                                            >TOTAL                     -           -          18.195        -
gremlin>
gremlin> g.V().has("name", "value0").limit(1000).profile()
==>Traversal Metrics
Step                                                               Count  Traversers       Time (ms)    % Dur
=============================================================================================================
JanusGraphStep([],[name.eq(value0)])                                                           0.609   100.00
    \_condition=(name = value0)
    \_orders=[]
    \_limit=1000
    \_isFitted=true
    \_isOrdered=true
    \_query=[(name = value0)](2000):name
    \_index=name
    \_index_impl=search
  optimization                                                                                 0.025
  optimization                                                                                 0.151
                                            >TOTAL                     -           -           0.609        -
gremlin>
gremlin> g.V().has("name", "value0").limit(100).profile()
==>Traversal Metrics
Step                                                               Count  Traversers       Time (ms)    % Dur
=============================================================================================================
JanusGraphStep([],[name.eq(value0)])                                 100         100          14.545   100.00
    \_condition=(name = value0)
    \_orders=[]
    \_limit=100
    \_isFitted=true
    \_isOrdered=true
    \_query=[(name = value0)](200):name
    \_index=name
    \_index_impl=search
  optimization                                                                                 0.015
  optimization                                                                                 0.155
  backend-query                                                                                0.000
    \_query=name:[(name = value0)](200):name
    \_limit=200
                                            >TOTAL                     -           -          14.545        -
```

The first query, g.V().has("name", "value0").limit(1000), hits the mixed index and response is cached.
The second query, g.V().has("name", "value0").limit(1000), hits the cache and thus no index query is fired.
The third query, g.V().has("name", "value0").limit(100), does not hit the cache because limit is also part of the key.

### PR DESCRIPTION
**Pull request by li-boxuan**
This refactors indexCache used in transaction which caches index query results.
Now indexCache supports subset query, i.e., cached query responses can be
leveraged if new query and cached query only differ in limit.

Closes #2194

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there an issue associated with this PR? Is it referenced in the commit message?
- [x] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](https://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?
- [ ] If this PR is a documentation-only change, have you added a `[doc only]`
  tag to the first line of your commit message to avoid spending CPU cycles in
  Travis CI when no code, tests, or build configuration are modified?

### Note:
Please ensure that once the PR is submitted, you check Travis CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by li-boxuan** [comment]
Test result (in comparison with 0.5.2 results in https://github.com/JanusGraph/janusgraph/issues/2194):

```
gremlin> graph = JanusGraphFactory.open("conf/janusgraph-cql-es.properties")
15:12:12 WARN  com.datastax.driver.core.NettyUtil  - Found Netty's native epoll transport, but not running on linux-based operating system. Using NIO instead.
==>standardjanusgraph[cql:[127.0.0.1]]
gremlin>
gremlin> mgmt = graph.openManagement()
==>org.janusgraph.graphdb.database.management.ManagementSystem@7282af25
gremlin> nameKey = mgmt.makePropertyKey('name').dataType(String.class).make()
==>name
gremlin> mgmt.buildIndex("name", Vertex.class).addKey(nameKey, Mapping.STRING.asParameter()).buildMixedIndex("search")
==>name
gremlin> mgmt.commit()
==>null
gremlin>
gremlin> g = graph.traversal()
==>graphtraversalsource[standardjanusgraph[cql:[127.0.0.1]], standard]
gremlin> for (i = 0; i < 10000; i++) {
......1> g.addV().property("name", "value" + i % 10).iterate();
......2> }
==>null
gremlin> graph.tx().commit()
==>null
gremlin>
gremlin> g.V().has("name", "value0").limit(1000).profile()
==>Traversal Metrics
Step                                                               Count  Traversers       Time (ms)    % Dur
=============================================================================================================
JanusGraphStep([],[name.eq(value0)])                                                          21.075   100.00
    \_condition=(name = value0)
    \_orders=[]
    \_limit=1000
    \_isFitted=true
    \_isOrdered=true
    \_query=[(name = value0)](2000):name
    \_index=name
    \_index_impl=search
  optimization                                                                                 0.042
  optimization                                                                                 1.743
  backend-query                                                        0                      14.674
    \_query=name:[(name = value0)](2000):name
    \_limit=2000
                                            >TOTAL                     -           -          21.075        -
gremlin>
gremlin> g.V().has("name", "value0").limit(1000).profile()
==>Traversal Metrics
Step                                                               Count  Traversers       Time (ms)    % Dur
=============================================================================================================
JanusGraphStep([],[name.eq(value0)])                                                           0.450   100.00
    \_condition=(name = value0)
    \_orders=[]
    \_limit=1000
    \_isFitted=true
    \_isOrdered=true
    \_query=[(name = value0)](2000):name
    \_index=name
    \_index_impl=search
  optimization                                                                                 0.031
  optimization                                                                                 0.123
                                            >TOTAL                     -           -           0.450        -
gremlin>
gremlin> g.V().has("name", "value0").limit(100).profile()
==>Traversal Metrics
Step                                                               Count  Traversers       Time (ms)    % Dur
=============================================================================================================
JanusGraphStep([],[name.eq(value0)])                                                           0.327   100.00
    \_condition=(name = value0)
    \_orders=[]
    \_limit=100
    \_isFitted=true
    \_isOrdered=true
    \_query=[(name = value0)](200):name
    \_index=name
    \_index_impl=search
  optimization                                                                                 0.016
  optimization                                                                                 0.105
                                            >TOTAL                     -           -           0.327        -
```

**Comment by rngcntr** [comment]
Hi @li-boxuan! I only skimmed your code and did not have a detailed look at it so far. But I see from your class names that you are using a Guava cache. There's an open issue (#2033) that has set the goal to remove the guava dependency sooner or later. Sadly, I don't know about alternatives but in terms of #2033, new code depending on guava is probably not welcome.

**Comment by li-boxuan** [comment]
@rngcntr Thanks for the reminder! Yes, I am also aware of that issue. An alternative is caffeine (https://github.com/ben-manes/caffeine) which is also used by Neo4j, Cassandra, ... but I never used it before.

The primary reason for still keeping usage of guava is because this PR is aimed at the improvement of the caching strategy only, without any possibly breaking change. It's almost transparent to users. If we switch to some other cache library other than guava (which was used in the old code), we would need extensive testing and evaluation as that would be a big move. We would probably substitute guava cache usage in other places at the same time.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/subquerycache/SubqueryCache.java
Comment (rngcntr) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/subquerycache/SubqueryCache.java:96]: Nitpick: Do `values` and `limit` have to be modifiable once their values are set by the constructor?

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/subquerycache/GuavaSubqueryCache.java
Comment (rngcntr) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/subquerycache/GuavaSubqueryCache.java:44]: The Java Stream API ist relatively inefficient in most simple cases. Instead you could use `List.subList()` here.

File: janusgraph-test/src/test/java/org/janusgraph/graphdb/query/QueryTest.java
Comment (rngcntr) [inline janusgraph-test/src/test/java/org/janusgraph/graphdb/query/QueryTest.java:136]: These tests do not cover the case in which we have a limit `x` first and then a limit `y > x` later. Otherwise the tests look good

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/subquerycache/SubqueryCache.java
Comment (rngcntr) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/subquerycache/SubqueryCache.java:46]: I like your verbose documentation :+1: 
But in this case, I think the description of the internal logic does not fit here. What you describe actually happens inside `GuavaSubqueryCache`, not `SubqueryCache`. So you should either move the description there ore move the caching logic here.
I would even prefer to have as much logic as possible right within `SubqueryCache` to ensure that all implementations are behaving logically equivalent.

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java
Comment (rngcntr) [inline janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java:2351]: If no limit is given, the cached result can be used if ordering is applied afterwards. But this wasn't done before so I'm also ok with declaring it out of scope of this PR.

**Review by rngcntr — COMMENTED** [review]
Overall, I like this optimization very much! But I've added a few comments.
In addition, I just thought about how the `range(start, end)` step would be handled. It could follow a similar pattern because `limit(n)` is actually just `range(0, n)`. Any ideas on that?

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/subquerycache/SubqueryCache.java
Comment (li-boxuan) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/subquerycache/SubqueryCache.java:96]: make sense, will remove redundant setters

File: janusgraph-test/src/test/java/org/janusgraph/graphdb/query/QueryTest.java
Comment (li-boxuan) [reply janusgraph-test/src/test/java/org/janusgraph/graphdb/query/QueryTest.java:136]: Agree. I think I will need to break this test case into multiple smaller test cases. Now several scenarios are mixed together and edge cases are more likely to be missing.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/subquerycache/SubqueryCache.java
Comment (li-boxuan) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/subquerycache/SubqueryCache.java:46]: agree, I will think over this and probably do some refactoring.

**Comment by li-boxuan** [comment]
Blocked by https://github.com/JanusGraph/janusgraph/issues/2217

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java
Comment (li-boxuan) [reply janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java]: The cached values are just vertex ids. Unless we also cache indexed field values (for mixed indexes, of course, since composite indexes do not support ordering query natively anyway), it's impossible to use caching results if ordering is applied afterward.

**Review by rngcntr — APPROVED** [review]

**Review by farodin91 — APPROVED** [review]
LGTM! Thank you @li-boxuan. Do you like to rebase before merging?

**Pull request merged**
